crates/uv-installer/src/compile.rs

```diff
@@ -0,0 +1,207 @@
+    let tempdir = tempdir().map_err(CompileError::TempFile)?;
+    let pip_compileall_py = tempdir.path().join("pip_compileall.py");
+    fs_err::tokio::write(&pip_compileall_py, COMPILEALL_SCRIPT)
+        .await
+        .map_err(CompileError::TempFile)?;
+    // We input the paths through stdin and get the successful paths returned through stdout.
+    let mut bytecode_compiler = Command::new(interpreter)
+        .arg(&pip_compileall_py)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::inherit())
+        .current_dir(dir)
+        // Otherwise stdout is buffered and we'll wait forever for a response
+        .env("PYTHONUNBUFFERED", "1")
+        .spawn()
+        .map_err(CompileError::PythonSubcommand)?;
+
+    // https://stackoverflow.com/questions/49218599/write-to-child-process-stdin-in-rust/49597789#comment120223107_49597789
+    // Unbuffered, we need to write immediately or the python process will get stuck waiting
+    let child_stdin = bytecode_compiler
+        .stdin
+        .take()
+        .expect("Child must have stdin");
+    let mut child_stdout = BufReader::new(
+        bytecode_compiler
+            .stdout
+            .take()
+            .expect("Child must have stdout"),
+    );
+
+    let result = worker_main_loop(receiver, child_stdin, &mut child_stdout).await;
+    // Avoid `/.venv/bin/python: can't open file` on stderr after success because the tempdir was
+    // dropped before the Python process even started.
+    let _ = bytecode_compiler.kill().await;
+    result
+}
+
+/// We use stdin/stdout as a sort of bounded channel. We write one path to stdin, then wait until
+/// we get the same path back from stdout. This way we ensure one worker is only working on one
+/// piece of work at the same time.
+async fn worker_main_loop(
+    receiver: Receiver<PathBuf>,
+    mut child_stdin: ChildStdin,
+    child_stdout: &mut BufReader<ChildStdout>,
+) -> Result<(), CompileError> {
+    let mut out_line = String::new();
+    while let Ok(source_file) = receiver.recv().await {
+        // Luckily, LF alone works on windows too
+        let bytes = format!("{}\n", source_file.display()).into_bytes();
+        // Ensure we don't get stuck here because some pipe ran full. The Python process isn't doing
+        // anything at this point except waiting for stdin, so this should be immediate.
+        let timeout = Duration::from_secs(1);
+        match tokio::time::timeout(timeout, child_stdin.write_all(&bytes)).await {
+            // Timeout
+            Err(_) => {
+                receiver.close();
+                return Err(CompileError::Timeout(timeout));
+            }
+            // Write error
+            Ok(Err(err)) => return Err(CompileError::ChildStdin(err)),
+            Ok(Ok(())) => {}
+        }
+
+        out_line.clear();
+        child_stdout
+            .read_line(&mut out_line)
+            .await
+            .map_err(CompileError::ChildStdout)?;
```

Comment:
If we have a timeout on reading from stdin, should we also have a timeout on writing to stdout?